# 0017-class-variable-overtaken.patch
#
# DE DONDE SALE. Parche 0011 de los 37 de syntax-transform/3.1/ del PR https://github.com/mkxp-z/mkxp-z/pull/304
# de mkxp-z, de white-axe (rama syntax-transform, 0be32ad119b054749cf3db8c72434f8fa1ac83ea).
# Ensenan al parser y al nucleo de Ruby 3.1 la gramatica y la semantica de
# Ruby 1.8, que es en lo que estan escritos cinco de los ocho juegos del censo.
#
# COMO SE APLICA. Sobre src/ruby-3.1.7 DESPUES de los parches 0001-0005 y en
# orden de numero: el 0006 (prelude) monta el mecanismo que los otros 36 usan,
# y varios de ellos tocan el mismo fichero, asi que el orden no es decorativo.
#
#     cd src/ruby-3.1.7
#     sed -i 's/\r$//' string.c
#     for n in $(seq 6 42); do
#       patch -p1 -i "../../patches/$(printf '%04d' $n)-"*.patch || break
#     done
#     rm -f parse.c id.h insns.inc insns_info.inc opt_sc.inc optinsn.inc \
#           optunifs.inc vm.inc vmtc.inc mjit_compile.inc
#
# Ese `rm` NO es opcional y cuesta una tarde descubrirlo. El tarball de Ruby
# trae parse.c y los .inc YA GENERADOS en el propio directorio de fuentes, y
# un `#include "insns.inc"` desde compile.c mira primero la carpeta del fichero
# que incluye: aunque el build regenere los suyos al lado del Makefile, gana el
# viejo. Con parse.c viejo los ocho parches de gramatica no llegan al binario y
# NADA lo dice; con los .inc viejos, al menos, no compila
# ('YARVINSN_mkxp_syntax_transform_raise_invalid_break' undeclared).
#
# El cuerpo del diff, de aqui abajo, es el del PR VERBATIM: los 37 aplicaron
# sin fuzz y sin rechazos sobre este arbol, asi que no habia nada que adaptar.
# Lo unico que hay que vigilar es string.c, que en este repositorio esta en
# CRLF y al que `patch` le deja el fichero entero en LF; se normaliza antes,
# para que ese cambio sea deliberado y no un efecto secundario.
#
# COMPROBADO. Con los 37 puestos, bison 3.8.2 acepta la gramatica (parse.y
# declara %expect 0: cualquier conflicto nuevo habria sido un error, y no hubo
# ninguno) y CRuby 3.1.7 cruza-compila para aarch64 con devkitA64 16.1.0 hasta
# libruby-static.a (44,7 MB), con los simbolos del mecanismo dentro
# (mkxp_syntax_transform_set_next_eval, mkxp_ec_is_syntax_transform_active,
# mkxp_syntax_transform_target_ruby_version_*) y 14 instrucciones nuevas en
# insns.inc. No se ha ejecutado: eso pide consola.
#
# QUE HACE FALTA ADEMAS. Estos parches por si solos NO cambian el
# comportamiento: mkxp_syntax_transform_target_ruby_version_* nace en -1 y la
# transformacion queda apagada, asi que el interprete se comporta como el de
# hoy. Encenderla es cosa del lado de mkxp-z (armar
# mkxp_syntax_transform_set_next_eval(1) alrededor del eval de cada seccion de
# Scripts.rxdata y fijar la version objetivo desde mkxp.json), que en este
# repositorio todavia no esta hecho.
#
# Licencia: GPL-2.0-or-later, la del PR. La cabecera original va debajo.
# ---------------------------------------------------------------------------

#
# This file is part of mkxp.
#
# Copyright (C) 2026 The mkxp-z authors
#
# mkxp is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# mkxp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with mkxp.  If not, see <http://www.gnu.org/licenses/>.

# In Ruby 1.8, if a class variable is defined in a derived class and then a
# class variable with the same name is defined in its base class, the value of
# the class variable as visible in the derived class is the original value that
# was set in the derived class. In Ruby versions later than 1.8 up to 2.7, the
# value is instead the new value that was set in the base class. In Ruby
# versions later than 2.7, attempting to read or write the class variable in
# the derived class raises a `RuntimeError`. Thus, after running this code
# snippet
#
#   class A
#   end
#   class B < A
#     @@a = 'B'
#   end
#   class A
#     @@a = 'A'
#   end
#   class B
#     p @@a
#   end
#
#  the value `"B"` will be printed in Ruby 1.8, the value `"A"` will be printed
#  in Ruby 1.9 through 2.7 and an error will occur on the `p @@a` line in Ruby
#  versions later than 2.7.
#
#  This code snippet, however, will print `"B"` in all Ruby versions without
#  any errors occurring.
#
#  class A
#    @@a = 'A'
#  end
#  class B < A
#    @@a = 'B'
#    p @@a
#  end
#
#  This code snippet will print `"b"` in Ruby 1.8, `"a"` in Ruby 1.9 through
#  Ruby 2.7 and throw an error in later versions on the line `@@a = 'b'`.
#
#  class A
#  end
#  class B < A
#    @@a = 'B'
#  end
#  class A
#    @@a = 'A'
#  end
#  class B
#    @@a = 'b'
#  end
#  class A
#    @@a = 'a'
#  end
#  class B
#    p @@a
#  end

--- a/variable.c
+++ b/variable.c
@@ -3458,12 +3458,24 @@ cvar_overtaken(VALUE front, VALUE target, ID id)
 	st_data_t did = (st_data_t)id;
 
         if (original_module(front) != original_module(target)) {
-            rb_raise(rb_eRuntimeError,
-                     "class variable % "PRIsVALUE" of %"PRIsVALUE" is overtaken by %"PRIsVALUE"",
-		       ID2SYM(id), rb_class_name(original_module(front)),
-		       rb_class_name(original_module(target)));
+            if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
+                rb_warning("class variable % "PRIsVALUE" of %"PRIsVALUE" is overridden by %"PRIsVALUE"",
+                           ID2SYM(id), rb_class_name(original_module(target)),
+                           rb_class_name(original_module(front)));
+            }
+            else if (mkxp_ec_is_syntax_transform_active(2, 7, -1)) {
+                rb_warn("class variable % "PRIsVALUE" of %"PRIsVALUE" is overtaken by %"PRIsVALUE"",
+                        ID2SYM(id), rb_class_name(original_module(front)),
+                        rb_class_name(original_module(target)));
+            }
+            else {
+                rb_raise(rb_eRuntimeError,
+                         "class variable % "PRIsVALUE" of %"PRIsVALUE" is overtaken by %"PRIsVALUE"",
+                         ID2SYM(id), rb_class_name(original_module(front)),
+                         rb_class_name(original_module(target)));
+            }
 	}
-	if (BUILTIN_TYPE(front) == T_CLASS) {
+	if (!mkxp_ec_is_syntax_transform_active(1, 8, -1) && BUILTIN_TYPE(front) == T_CLASS) {
 	    st_delete(RCLASS_IV_TBL(front), &did, 0);
 	}
     }
@@ -3472,21 +3484,28 @@ cvar_overtaken(VALUE front, VALUE target, ID id)
 static VALUE
 find_cvar(VALUE klass, VALUE * front, VALUE * target, ID id)
 {
-    VALUE v = Qundef;
+    VALUE v = Qundef, _v;
+    bool syntax_transform = mkxp_ec_is_syntax_transform_active(1, 8, -1);
     CVAR_ACCESSOR_SHOULD_BE_MAIN_RACTOR();
-    if (cvar_lookup_at(klass, id, (&v))) {
+    if (cvar_lookup_at(klass, id, (&_v))) {
         if (!*front) {
             *front = klass;
         }
         *target = klass;
+        if (!syntax_transform || v == Qundef) {
+            v = _v;
+        }
     }
 
     for (klass = cvar_front_klass(klass); klass; klass = RCLASS_SUPER(klass)) {
-	if (cvar_lookup_at(klass, id, (&v))) {
+	if (cvar_lookup_at(klass, id, (&_v))) {
             if (!*front) {
                 *front = klass;
             }
             *target = klass;
+            if (!syntax_transform || v == Qundef) {
+                v = _v;
+            }
         }
     }
 
@@ -3529,6 +3548,9 @@ rb_cvar_set(VALUE klass, ID id, VALUE val)
     CVAR_LOOKUP(0, {if (!front) front = klass; target = klass;});
     if (target) {
 	cvar_overtaken(front, target, id);
+        if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
+            target = front;
+        }
     }
     else {
 	target = tmp;
